File: week2/cpp/sharpen.cpp
/*
Copyright 2017 BIG VISION LLC ALL RIGHTS RESERVED

This code is made available to the students of 
the online course titled "Computer Vision for Faces" 
by Satya Mallick for personal non-commercial use. 

Sharing this code is strictly prohibited without written
permission from Big Vision LLC. 

For licensing and other inquiries, please email 
spmallick@bigvisionllc.com 
*/

#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>
#include <string>

using namespace cv;
using namespace std;

int main(int argc, char ** argv)
{
  // Read the image
  string filename = "../data/images/mountain.jpeg";
  if (argc == 2)
  {   
    filename = argv[1];
  }
  
  Mat image = imread(filename);
  Mat output,output2;
  // Sharpen kernel
  Mat sharpen = (Mat_<double>(3,3) << 0, -1, 0, -1, 5, -1, 0, -1, 0);
  if(!image.data )
    return -1;

  // Use 2D filter using the sharpening kernel
  // - 1 is used to keep the depth of the resultant image same as the source image
  filter2D(image, output, -1, sharpen);

  Mat combined;
  cv::hconcat(image, output, combined);
  namedWindow("Original Image -- Sharpening Result",CV_WINDOW_AUTOSIZE);
  imshow("Original Image -- Sharpening Result",combined);  
  waitKey(0);
  destroyAllWindows();
  
  imwrite("results/Sharpening.jpg",output);
}

File: week2/python/sharpen.py
# Copyright 2017 BIG VISION LLC ALL RIGHTS RESERVED
#
# This code is made available to the students of
# the online course titled "Computer Vision for Faces"
# by Satya Mallick for personal non-commercial use.
#
# Sharing this code is strictly prohibited without written
# permission from Big Vision LLC.
#
# For licensing and other inquiries, please email
# spmallick@bigvisionllc.com
#

import cv2,argparse
import numpy as np

ap = argparse.ArgumentParser()
ap.add_argument("-f", "--filename")
args = vars(ap.parse_args())

filename = "../data/images/mountain.jpeg"
if args['filename']:
	filename = args['filename']

image = cv2.imread(filename)

if image is None:
  	print("Image not read")

# cv2.imshow("Original", image)

# Sharpen kernel
sharpen = np.array((
	[0, -1, 0],
	[-1, 5, -1],
	[0, -1, 0]), dtype="int")

# Using 2D filter by applying the sharpening kernel
sharpenOutput = cv2.filter2D(image, -1, sharpen)

combined = np.hstack([image, sharpenOutput])

img_name = "Original Image -- Sharpening Result"
cv2.namedWindow(img_name, cv2.WINDOW_NORMAL | cv2.WINDOW_KEEPRATIO)
cv2.resizeWindow(img_name, 800,800)
cv2.imshow(img_name, combined)

cv2.waitKey(0)
cv2.destroyAllWindows()
cv2.imwrite("results/sharpening.jpg",sharpenOutput)
